app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js

```diff
@@ -0,0 +1,209 @@
+import { ethErrors } from 'eth-rpc-errors'
+import validUrl from 'valid-url'
+import { omit } from 'lodash'
+import { MESSAGE_TYPE } from '../../../../../shared/constants/app'
+import {
+  isPrefixedFormattedHexString,
+  isSafeChainId,
+} from '../../../../../shared/modules/network.utils'
+import { jsonRpcRequest } from '../../../../../shared/modules/rpc.utils'
+import { CHAIN_ID_TO_NETWORK_ID_MAP } from '../../../../../shared/constants/network'
+
+const addEthereumChain = {
+  methodNames: [MESSAGE_TYPE.ADD_ETHEREUM_CHAIN],
+  implementation: addEthereumChainHandler,
+}
+export default addEthereumChain
+
+async function addEthereumChainHandler(
+  req,
+  res,
+  _next,
+  end,
+  { addCustomRpc, customRpcExistsWith, requestUserApproval },
+) {
+  if (!req.params?.[0] || typeof req.params[0] !== 'object') {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected single, object parameter. Received:\n${req.params}`,
+      }),
+    )
+  }
+
+  const { origin } = req
+
+  const {
+    chainId,
+    chainName = null,
+    blockExplorerUrls = null,
+    nativeCurrency = null,
+    rpcUrls,
+  } = req.params[0]
+
+  const otherKeys = Object.keys(
+    omit(req.params[0], [
+      'chainId',
+      'chainName',
+      'blockExplorerUrls',
+      'iconUrl',
```

Comment:
I just remembered: It was discussed on EIP-3085 that this field should be changed to `iconUrls` (plural) and be string array. I'm going to update the EIP to encompass that.

We don't currently do anything with this property anyway as far as I can tell, but just, FYI.